### DISCUSSION TITLE
What should the Python support level going forward be in html5lib?

### DISCUSSION THREAD

**Discussion by ambv** [html5lib/html5lib-python#561] [Polls]
Python 2.7 has been EOL since April 2020.
Python 3.5 has been EOL since September 2020.
Python 3.6 has been EOL since September 2021.
Python 3.7 is going away in June 2023.

Python 3.5 and 3.6 are no longer available on GitHub Actions for CI purposes.

Dropping Python 2 would allow us to stop worrying about narrow builds and use `pathlib`. Adopting "3.6+" would allow us to use f-strings. Adopting "3.7+" would let us cover `html5lib` with typing annotations and enable type checking.

**Comment by gsnedders** [comment]
FWIW: historically I believe the baseline was "whatever pip requires" plus "whatever web-platform-tests requires" (the latter mostly because… it was as a dependency of that which essentially got maintenance work done), where the former seems to be 3.7 and the latter seems to be 3.6 for now (mostly due to the Firefox tooling: https://bugzilla.mozilla.org/show_bug.cgi?id=1734402, because Ubuntu and other distros want to be able to build with their packaged and maintained Python version)

**Reply by ambv** [reply]
Two additional advantages of going 3.7+ would be that we could use `dataclasses` and with type annotations and dataclasses we could adopt `mypyc` to compile the library. Black compiled with mypyc goes 2X as fast as vanilla Python. That's better than what Cython can give us because mypyc supports a subset of Python with some shortcuts in object attribute access, among others.

**Reply by ambv** [reply]
Another advantage: we could drop AppVeyor entirely and rely on just GitHub Actions which is much faster and better integrated with GitHub.

**Reply by gsnedders** [reply]
Given the Mozilla bug just got closed with them moving to Python 3.7 (presumably as ToT is going to release _after_ Ubuntu 18.04 LTS goes out of LTS support)… I guess that makes this clearly 3.7 now?

I do agree with need to reach a point of 3.7, the only question was quite how soon we could do that—though it clearly wasn't far off.

**Comment by ambv** [comment]
OK. The consensus is to release 1.2 with support for 2.7 - 3.11, and immediately after cut support for 2.7, 3.5, and 3.6.

**Comment by satch421** [comment]
I have been having trouble with the Django Rest framework dependency on html5lib for some time.  Although I was apparently able to install "sudo apt -y install python3-html5lib", I still get an import error when I attempt to install restframework: "cannot import name 'html5lib' from 'pip._vendor' (/usr/lib/python3/dist-packages/pip/_vendor/__init__.py)".  Any thoughts or suggestions for me?  thx,  pete.godston@gmail.com

**Comment by mbron** [comment]
Hi, hope everyone is doing well. Since March things have moved on and python 3.7 has ended security updates last month. Raising a couple of points for discussion:
* Are there any thoughts on a timeline for the release of the 1.2 and beyond versions? 
* Will any (security) issues raised before that be considered in a patch release until that time?

Many thanks and keep up the good work!
